compiler/curse_interpreter/iter.curse

```diff
@@ -0,0 +1,50 @@
+dynamic_import "std.curse"
+
+struct Fn |Lhs * Rhs * Out| {}
+
+struct Iterator |St * T| {
+    state: St,
+    next_fn: Fn (St * {} * { next_value: Option T, next_state: St }),
+}
+
+fn range |begin, end| {
+    state: { begin, end },
+    next_fn: |{ begin, end }| begin < end then begin
+        in |next_value| {
+            next_value,
+            next_state: { begin: begin + 1, end },
+        },
+}
+
+fn map |iter, map_fn| {
+    state: iter,
+    next_fn: |iter|
+        next of iter
+        in |{ iter, next_value }| {
+            next_value: next_value option_map map_fn,
+            next_state: iter,
+        },
+}
+
+// Takes an iterator and returns a record containing:
+// 1. the updated iterator,
+// 2. the optional value produced
+fn next |{ state, next_fn }|
+    next_fn of state in |{ next_value, next_state }| {
+        iter: { state: next_state, next_fn },
+        next_value,
+    }
+
+fn foldr |iter, fold|
+    next of iter in |{ iter, next_value }|
+        next_value
+            option_map (|value| iter foldr fold fold value)
```

Comment:
I have a strong love-hate relationship with this line...

Reply:
I totally agree. It took me a second to parse, `iter foldr fold fold value` is a wild line, but is still way more readable than the `fold(foldr(iter, fold), value)` you'd get in a more conventional language.

One thing this makes me realize that I like about curse though, is how every binary operator has the same associativity and precedence. When faced with a tough to parse line like this one you can just read it left to right and you're guaranteed it will work out. Haskell uses tons of custom binary operators, but you can define custom associativity and precedences for them, which means I never have any idea when you need parens. E.g. you could do something like
```haskell
iter `foldr` fold `fold` value
```
in Haskell (predenting `foldr`'s arguments go the other way), but you might or might not need parens depending on if the associativity and precedences work out. And its really hard to remember the precedences and associativities of every binary operator. Plus you need to keep switching between reading right to left and left to right based on whether you're using `.` or `>>=`! All this makes Haskell so much harder to read, for me at least, than curse. In curse you just read it left to right and that's it. it's so nice.